extensions/wikidata/src/org/openrefine/wikidata/commands/ConnectionManager.java

```diff
@@ -0,0 +1,219 @@
+/*******************************************************************************
+ * MIT License
+ *
+ * Copyright (c) 2018 Antonin Delpeuch
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ ******************************************************************************/
+package org.openrefine.wikidata.commands;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.refine.ProjectManager;
+import com.google.refine.preference.PreferenceStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wikidata.wdtk.wikibaseapi.ApiConnection;
+import org.wikidata.wdtk.wikibaseapi.BasicApiConnection;
+import org.wikidata.wdtk.wikibaseapi.LoginFailedException;
+import org.wikidata.wdtk.wikibaseapi.OAuthApiConnection;
+import org.wikidata.wdtk.wikibaseapi.apierrors.MediaWikiApiErrorException;
+
+import javax.servlet.http.Cookie;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Manages a connection to Wikidata.
+ * <p>
+ * The connection can be either {@link BasicApiConnection} or {@link OAuthApiConnection}.
+ * <p>
+ * This class is also hard-coded for Wikidata,
+ * it will be generalized to other Wikibase instances soon.
+ *
+ * @author Antonin Delpeuch
+ * @author Lu Liu
+ */
+
+public class ConnectionManager {
+
+    final static Logger logger = LoggerFactory.getLogger("connection_manager");
+
+    /**
+     * We used this key to read/write credentials from/to preferences in the past, which is insecure.
+     * Now this key is kept only to delete those credentials in the preferences.
+     */
+    public static final String PREFERENCE_STORE_KEY = "wikidata_credentials";
+
+    public static final int CONNECT_TIMEOUT = 5000;
```

Comment:
This seems like it might be too short, but I guess we can find out when users test.